### DISCUSSION TITLE
লেয়াউট ইডিটর / বিল্ডার

### DISCUSSION THREAD

**Discussion by qomarhsn** [OpenBangla/OpenBangla-Keyboard#345] [Ideas]
লিনাক্সে শিফট হওয়ার পর থেকে ওপেন বাংলা কীবোর্ড ব্যবহার করতেছি। এত সুন্দর করে লিনাক্সে বাংলা লেখার উপযোগী একটা টুল বানানোর জন্য ওপেন বাংলা টিমকে অসংখ্য ধন্যবাদ। সফটওয়্যারটিতে ইন্সটল লেয়াউট নামে একটা অপশন আছে। আমি জানতে চাচ্ছিলাম! এখানে ইন্সটল করার জন্য লেয়াউট কোথা থেকে পাব? আর সফটওয়্যারটিতে বিল্টিন একটা লেয়াউট ইডিটর বা লেয়াউট বিল্ডার অপশন যুক্ত করে দিলে ভালো হয়।  আর বর্তমানে লায়াউট বিল্ড করার প্রসেসটা কী?
![Screenshot from 2023-08-12 01-33-28](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/142052078/a673fa5e-b76d-4cb8-8a22-d0f02e4a7de3)

**Comment by mominul** [comment]
@kamrulhasanio Thanks! You can use https://github.com/NabilSnigdho/obk-layout-editor for building new layouts. Here is the [download link](https://github.com/NabilSnigdho/obk-layout-editor/releases/tag/obk-layout-editor-v0.0.2).

**Comment by qomarhsn** [comment]
![Screenshot from 2023-08-12 02-06-33](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/142052078/a276b093-7c58-42dd-b9bc-e532ea7c8dfa)
এই সফটওয়্যার এ বাংলা ইনপুট নিচ্ছে না। আমি অভ্র ফোনেটিক দিয়ে চেষ্টা করেছি। তবে কোথাও থেকে কপি করে দিলে নিচ্ছে।

**Reply by mominul** [reply]
@NabilSnigdho please look into this issue

**Reply by NabilSnigdho** [reply]
Unfortunately the fcitx and other IMEs doesn't work on the AppImage build of obk-layout-editor.

I might make build releases for other linux distributions soon.
For now, you can just extract & run the `obk-layout-editor` executable from the `.deb` file.

**Reply by NabilSnigdho** [reply]
@kamrulhasanio here's a `PKGBUILD` to install the package in archlinux
```sh
pkgname="obk-layout-editor"
pkgver="0.0.2"
pkgrel=1
pkgdesc="Utility to create/edit OpenBangla Keyboard layout files."
arch=('x86_64')
url="https://github.com/NabilSnigdho/obk-layout-editor"
license=('MPL 2')
source=("https://github.com/NabilSnigdho/obk-layout-editor/releases/download/obk-layout-editor-v$pkgver/obk-layout-editor_${pkgver}_amd64.deb")
sha512sums=("SKIP")
depends=(
  'libwebkit2gtk-4.0.so'
  'libgtk-3.so'
)

package(){
  tar -xz -f data.tar.gz -C "${pkgdir}"
}
```

**Reply by qomarhsn** [reply]
[Screencast from 2023-08-26 16-17-43.webm](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/142052078/f9636ca7-dc22-4a39-bb2d-f31a5d8afa87)
এটা ওপেনই হচ্ছে না।

**Reply by NabilSnigdho** [reply]
@kamrulhasanio can you check the output at the terminal of command `obk-layout-editor`? also did use the PKGBUILD?

**Reply by qomarhsn** [reply]
Following your suggestion, I attempted the command "**obk-layout-editor**" and encountered the following output: "**obk-layout-editor: error while loading shared libraries: libwebkit2gtk-4.0.so.37: cannot open shared object file: No such file or directory.**" Subsequently, I resolved the issue by installing "**webkit2gtk**." I am extremely grateful for your assistance; many, many thanks to you! My problem has now been successfully resolved.

**Reply by NabilSnigdho** [reply]
Oh, I had forgotten about the dependencies, just updated the PKGBUILD above.

**Reply by qomarhsn** [reply]
One more thing, after constructing the layout, it designates a default save location where the layout can only be saved. Regrettably, it's not feasible to save it in a different location and subsequently install it using the "layout install" option found within the keyboard settings. What options are available to address this issue? I aspire to share my created layout with others; however, it appears that the only method is to reopen the layout builder and save it directly into the "/home/user/.local/share/openbangla-keyboard/layouts/" directory. Regrettably, there isn't an alternative approach to installing the layout without transmitting the file to this particular directory.

**Reply by NabilSnigdho** [reply]
@kamrulhasanio well you can share the `*.json` file with others and tell them to place it in `~/.local/share/openbangla-keyboard/layouts/`. There's no need to reopen the file in OBKLE to install it.

@mominul shouldn't there be an option to import OBK layout *.json files in the UI? I might create a PR in that case.

**Reply by qomarhsn** [reply]
![Screenshot from 2023-08-27 14-30-22](https://github.com/OpenBangla/OpenBangla-Keyboard/assets/142052078/2f2c1644-c35d-408a-b333-ef3f166f61b8)
There is an option called Install Layout!
But it's not working.

**Reply by mominul** [reply]
> @mominul shouldn't there be an option to import OBK layout *.json files in the UI? I might create a PR in that case.

@NabilSnigdho yeah, that would be nice!

**Reply by mominul** [reply]
> There is an option called Install Layout!
But it's not working.

@kamrulhasanio it currently only supports Avro Keyboard 5.0 layout files, as when I implemented it the layout editor was not available. But as we now have a layout editor, the next step would be extend this to accept the generated `.json` files.

**Reply by qomarhsn** [reply]
Waiting for the next update

**Reply by mominul** [reply]
@kamrulhasanio It has been implemented on the develop branch, have you tried it out yet?

**Reply by qomarhsn** [reply]
🎉 @mominul! It's working splendidly! Many thanks for flawlessly implementing this feature. 🙌

**Reply by mominul** [reply]
I'm glad it's useful to you! :blush: The thanks should go to @NabilSnigdho who developed the editor and wrote the integration code!
